yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,315 @@
+                    dim,
+                    *grid_limit,
+                )
+        if len(grid_limit) != 2:
+            raise ValueError(
+                f"grid_limit_{dim} must have 2 dimensions, but it has {len(grid_limit)}"
+            )
+
+        return grid_limit
+
+    def _validate_grid_shape(
+        self, grid_shape: Optional[Tuple[int, int, int]] = None
+    ) -> Tuple[int, int, int]:
+        if grid_shape is None:
+            grid_shape = (100, 100, 100)
+            mylog.info(
+                "grid_shape not provided, using (nz, ny, nx) = (%i, %i, %i)",
+                *grid_shape,
+            )
+        if len(grid_shape) != 3:
+            raise ValueError(
+                f"grid_shape must have 3 dimensions, but it has {len(grid_shape)}"
+            )
+        return grid_shape
+
+    def _round_grid_guess(self, bounds: Tuple[float, float], unit_str: str):
+        # rounds the bounds to the closest 10 km increment that still contains
+        # the grid_limit
+        for findstr, repstr in self._field_info_class.unit_subs:
+            unit_str = unit_str.replace(findstr, repstr)
+        limits = unyt_array(bounds, unit_str).to("km")
+        limits[0] = np.floor(limits[0] / 10.0) * 10.0
+        limits[1] = np.ceil(limits[1] / 10.0) * 10.0
+        return tuple(limits.to(unit_str).tolist())
+
+    def _set_code_unit_attributes(self):
+        with self._handle() as xr_ds_handle:
+            length_unit = xr_ds_handle.variables["x"].attrs["units"]
+        self.length_unit = self.quan(1.0, length_unit)
+        self.mass_unit = self.quan(1.0, "kg")
+        self.time_unit = self.quan(1.0, "s")
+
+    def _parse_parameter_file(self):
+        self.parameters = {}
+
+        with self._handle() as xr_ds_handle:
+            x, y, z = (xr_ds_handle.coords[d] for d in "xyz")
+
+            self.origin_latitude = xr_ds_handle.origin_latitude[0]
+            self.origin_longitude = xr_ds_handle.origin_longitude[0]
+
+            self.domain_left_edge = np.array([x.min(), y.min(), z.min()])
+            self.domain_right_edge = np.array([x.max(), y.max(), z.max()])
+            self.dimensionality = 3
+            dims = [xr_ds_handle.dims[d] for d in "xyz"]
+            self.domain_dimensions = np.array(dims, dtype="int64")
+            self._periodicity = (False, False, False)
+            self.current_time = float(xr_ds_handle.time.values)
+
+        # Cosmological information set to zero (not in space).
+        self.cosmological_simulation = 0
+        self.current_redshift = 0.0
+        self.omega_lambda = 0.0
+        self.omega_matter = 0.0
+        self.hubble_constant = 0.0
+
+    @classmethod
+    def _is_valid(cls, filename, *args, **kwargs):
+        # This accepts a filename or a set of arguments and returns True or
+        # False depending on if the file is of the type requested.
+
+        try:
+            ds = xr.open_dataset(filename, engine="netcdf4")
```

Comment:
Should we explicitly close this?

Reply:
Are you thinking something along the lines of such?:

```
if hasattr(ds, "attrs") and isinstance(ds.attrs, dict):
    con = "Conventions"
    conv_check = "CF/Radial" in ds.attrs.get(con, "") + ds.attrs.get(con.lower(), "")
    ds.close()
    return conv_check
ds.close()
return False
```

Reply:
yes, I think we should... or change it to use `xr.open_dataset` as a context manager. I'll push up a change shortly.

Reply:
just pushed up a change (didn't see your comment til too late, @zssherman!). I ended up switching to `with xr.open_dataset(filename, engine="netcdf4") as ds:`  so that it'd be closed and shuffling things around to make that work.

Reply:
No worries! 'with' is better to use, don't know why I didn't think of that first.